### PR TITLE
Add a search context to search queries.

### PR DESCRIPTION
**Pull request by fmassot**
### Description

I think it would be nice to remove all these once cells for a bunch of parameters, caching and semaphores.

What do you think?

I need to clean a bit the code but you will get the idea.

### How was this PR tested?

`make test-all` passing except for kafka but I need to check if I have the same issue on main.

### CODE REVIEW

File: quickwit-config/src/config.rs [hunk 1]
```diff
@@ -115,12 +114,6 @@ impl Default for IndexerConfig {
     }
 }
 
-pub static SEARCHER_CONFIG_INSTANCE: once_cell::sync::OnceCell<SearcherConfig> = OnceCell::new();
```
Comment (fmassot) [inline quickwit-config/src/config.rs:118]: It's quite strange to see this instance in `quickwit-config`

File: quickwit-search/src/fetch_docs.rs [hunk 2]
```diff
@@ -29,15 +29,16 @@ use tantivy::{IndexReader, ReloadPolicy};
...
 use crate::GlobalDocAddress;
 
 /// Given a list of global doc address, fetches all the documents and
 /// returns them as a hashmap.
-#[allow(clippy::needless_lifetimes)]
 async fn fetch_docs_to_map(
     mut global_doc_addrs: Vec<GlobalDocAddress>,
     index_storage: Arc<dyn Storage>,
     splits: &[SplitIdAndFooterOffsets],
+    searcher_context: Arc<SearcherContext>,
```
Comment (guilload) [inline quickwit-search/src/fetch_docs.rs:41]: Just like `self`, I'd pass the context argument first in all those functions.

File: quickwit-search/src/lib.rs [hunk 3]
```diff
@@ -264,9 +267,6 @@ pub async fn start_searcher_service(
     storage_uri_resolver: StorageUriResolver,
     cluster: Arc<Cluster>,
 ) -> anyhow::Result<Arc<dyn SearchService>> {
-    SEARCHER_CONFIG_INSTANCE
```
Comment (fmassot) [inline quickwit-search/src/lib.rs:267]: This prevents from running several servers in a given test.

File: quickwit-search/src/search_stream/leaf.rs [hunk 4]
```diff
@@ -100,32 +84,39 @@ async fn leaf_search_results_stream(
     storage: Arc<dyn Storage>,
     splits: Vec<SplitIdAndFooterOffsets>,
     doc_mapper: Arc<dyn DocMapper>,
+    searcher_context: Arc<SearcherContext>,
 ) -> impl futures::Stream<Item = crate::Result<LeafSearchStreamResponse>> + Sync + Send + 'static {
-    let max_num_concurrent_split_streams = get_max_num_concurrent_split_streams();
+    let max_num_concurrent_split_streams = 10;
```
Comment (guilload) [inline quickwit-search/src/search_stream/leaf.rs:89]: Set as value of the context?

File: quickwit-search/src/search_stream/leaf.rs [hunk 5]
```diff
@@ -100,32 +84,39 @@ async fn leaf_search_results_stream(
...
     stream_request: SearchStreamRequest,
     storage: Arc<dyn Storage>,
+    searcher_context: Arc<SearcherContext>,
 ) -> crate::Result<LeafSearchStreamResponse> {
-    let _leaf_split_stream_permit = get_split_stream_permit().await;
+    let _leaf_split_stream_permit = searcher_context
+        .split_stream_semaphore
+        .acquire()
+        .await
+        .expect("blbhbbh");
```
Comment (guilload) [inline quickwit-search/src/search_stream/leaf.rs:117]: Fix.

File: quickwit-storage/src/cache/quickwit_cache.rs [hunk 6]
```diff
@@ -40,29 +42,28 @@ impl From<Vec<(&'static str, Arc<dyn Cache>)>> for QuickwitCache {
...
     }
-}
 
-impl QuickwitCache {
+    /// Drops all the cache.
     pub fn empty() -> QuickwitCache {
         QuickwitCache::from(Vec::new())
     }
 
+    /// Drops all the cache.
```
Comment (guilload) [inline quickwit-storage/src/cache/quickwit_cache.rs:66]: Remove?

File: quickwit-storage/src/cache/quickwit_cache.rs [hunk 7]
```diff
@@ -22,15 +22,17 @@ use std::path::{Path, PathBuf};
...
 
 use crate::cache::{Cache, MemorySizedCache};
 use crate::metrics::CacheMetrics;
 use crate::OwnedBytes;
 
 const FULL_SLICE: Range<usize> = 0..usize::MAX;
 
-pub(crate) struct QuickwitCache {
+/// Quickwit long term storage cache.
+/// It is currently used by the search service to cache fast field only.
```
Comment (guilload) [inline quickwit-storage/src/cache/quickwit_cache.rs:33]:
```suggestion
/// It is currently used by the search service to cache fast fields only.
```

**Review by guilload — APPROVED** [review]

File: quickwit-search/src/fetch_docs.rs (hunk 2 above)
Comment (fmassot) [reply quickwit-search/src/fetch_docs.rs]: sounds a good idea yes

File: quickwit-search/src/search_stream/leaf.rs (hunk 4 above)
Comment (fmassot) [reply quickwit-search/src/search_stream/leaf.rs]: yes, fixed.

**Pull request merged**